src/tools/dtrust-tool.c

```diff
@@ -262,50 +857,254 @@ main(int argc, char *argv[])
+		pin_verify = pin_change;
+		if (pin_change < 0)
+			goto out;
+	}
+
+	if (opt_verify != NULL) {
+		pin_verify = parse_pin(card, opt_verify, "Verification", &opt_can_verify);
+		if (pin_verify < 0)
+			goto out;
+	}
+
+	if (opt_resume != NULL) {
+		opt_can_verify = 1;
+		pin_resume = parse_pin(card, opt_resume, "Resume", NULL);
+		if (pin_resume < 0)
+			goto out;
+	}
+
+	if (opt_unblock != NULL) {
+		pin_unblock = parse_pin(card, opt_unblock, "Unblock", NULL);
+		if (pin_unblock < 0)
+			goto out;
+	}
+
+	/* D-Trust Card 5 requires PACE authentication with CAN */
+	if (opt_can_verify &&
+			card->type >= SC_CARD_TYPE_DTRUST_V5_1_STD &&
+			card->type <= SC_CARD_TYPE_DTRUST_V5_4_MULTI) {
+		struct sc_pin_cmd_data data;
+
+		memset(&data, 0, sizeof(data));
+		data.cmd = SC_PIN_CMD_VERIFY;
+		data.pin_type = SC_AC_CHV;
+		data.pin_reference = PACE_PIN_ID_CAN;
+
+		if (opt_can_enter) {
+			r = get_pin(&can, "CAN", 0);
+			if (r < 0)
+				goto out;
+
+			data.pin1.data = (const unsigned char *)can;
+			data.pin1.len = strlen(can);
+		}
+
+		r = sc_select_file(card, sc_get_mf_path(), NULL);
+		if (r)
+			goto out;
+
+		r = sc_pin_cmd(card, &data, NULL);
+		if (r) {
+			fprintf(stderr, "Error verifying CAN.\n");
+			goto out;
+		}
+	}
 
 	if (opt_status) {
-		if (card->type == SC_CARD_TYPE_DTRUST_V4_1_STD ||
-				card->type == SC_CARD_TYPE_DTRUST_V4_1_MULTI ||
-				card->type == SC_CARD_TYPE_DTRUST_V4_1_M100)
-			pin_status(card, 0x03, "Card Holder PIN");
-		pin_status(card, 0x04, "Card Holder PUK");
-		pin_status(card, 0x87, "Signature PIN");
-		pin_status(card, 0x8B, "Transport PIN");
+		switch (card->type) {
+		case SC_CARD_TYPE_DTRUST_V4_1_STD:
+		case SC_CARD_TYPE_DTRUST_V4_1_MULTI:
+		case SC_CARD_TYPE_DTRUST_V4_1_M100:
+			pin_status(card, DTRUST4_PIN_ID_PIN_CH, "Card Holder PIN", 0);
+			/* FALLTHRU */
+
+		case SC_CARD_TYPE_DTRUST_V4_4_STD:
+		case SC_CARD_TYPE_DTRUST_V4_4_MULTI:
+			/* We have to select the QES app to verify and change the Signature PIN. */
+			sc_format_path("3F000101", &path);
+			r = sc_select_file(card, &path, NULL);
+			if (r)
+				goto out;
+
+			pin_status(card, DTRUST4_PIN_ID_PUK_CH, "Card Holder PUK", 0);
+			pin_status(card, DTRUST4_PIN_ID_QES, "Signature PIN", 0);
+
+			/* According to the spec, the local bit has to be set. */
+			pin_status(card, 0x80 | DTRUST4_PIN_ID_PIN_T, "Transport PIN", 1);
+			break;
+
+		case SC_CARD_TYPE_DTRUST_V5_1_STD:
+		case SC_CARD_TYPE_DTRUST_V5_1_MULTI:
+		case SC_CARD_TYPE_DTRUST_V5_1_M100:
+			r = sc_select_file(card, sc_get_mf_path(), NULL);
+			if (r)
+				goto out;
+
+			pin_status(card, DTRUST5_PIN_ID_PIN_T_AUT, "Transport PIN (Authentication)", 1);
+
+			/* We have to select the eSign app to verify and change the Authentication PIN. */
+			sc_format_path("3F000102", &path);
+			r = sc_select_file(card, &path, NULL);
+			if (r)
+				goto out;
+
+			pin_status(card, DTRUST5_PIN_ID_AUT, "Authentication PIN", 0);
+			/* FALLTHRU */
```

Comment:
fall through is recognized as well and it is gramatically correct

Reply:
fixed